models/trie_mimc/frontend_proof.go

```diff
@@ -0,0 +1,477 @@
+package trie_mimc
+
+import (
+	"github.com/consensys/gnark/frontend"
+	"github.com/consensys/gnark/std/hash/mimc"
+)
+
+// Return the result of right shift by 1, input size = 32
+// Input: {42, 194, 10, 96, 133, 113, 88, 31, 86, 136, 60, 65, 11, 106, 226, 218, 169, 220, 186, 36, 114, 230, 53, 147, 171, 202, 12, 106, 45, 89, 231, 132}
+// Output: {42, 194, 10, 96, 133, 113, 88, 31, 86, 136, 60, 65, 11, 106, 226, 218, 169, 220, 186, 36, 114, 230, 53, 147, 171, 202, 12, 106, 45, 89, 231}
+func rightShift1(api frontend.API, input frontend.Variable) frontend.Variable {
+	var lsb8 frontend.Variable = 0
+	var multiplier frontend.Variable = 1
+	inputBinary := api.ToBinary(input)
+	for i := 0; i < 8; i++ {
+		lsb8 = api.Add(lsb8, api.Mul(inputBinary[i], multiplier))
+		multiplier = api.Mul(multiplier, 2)
+	}
+	return api.DivUnchecked(api.Sub(input, lsb8), 256)
+}
+
+func leastNBytes(api frontend.API, input frontend.Variable, N int) frontend.Variable {
+	var lsb frontend.Variable = 0
+	var multiplier frontend.Variable = 1
+	inputBinary := api.ToBinary(input)
+	if N >= 32 {
+		return api.Add(input, 0)
+	}
+	for i := 0; i < 8*N; i++ {
+		lsb = api.Add(lsb, api.Mul(inputBinary[i], multiplier))
+		multiplier = api.Mul(multiplier, 2)
+	}
+	return lsb
+}
+
+// Right shift N Bytes
+func rightShiftNBytes(api frontend.API, input frontend.Variable, N int) frontend.Variable {
+	var lsb frontend.Variable = 0
+	var multiplier frontend.Variable = 1
+	inputBinary := api.ToBinary(input)
+	if N >= 32 {
+		return 0
+	}
+	for i := 0; i < N*8; i++ {
+		lsb = api.Add(lsb, api.Mul(inputBinary[i], multiplier))
+		multiplier = api.Mul(multiplier, 2)
+	}
+
+	var divider frontend.Variable = 1
+	for i := 0; i < N; i++ {
+		divider = api.Mul(divider, 256)
+	}
+	return api.DivUnchecked(api.Sub(input, lsb), divider)
+}
+
+func hashVectors(api frontend.API, hFunc mimc.MiMC,
+	hashes ...frontend.Variable) frontend.Variable {
+
+	hFunc.Write(hashes[0])
+	hFunc.Write(rightShift1(api, hashes[1]))
+	for n, h := range hashes[2:] {
+		hFunc.Write(api.Add(NBytesLeftShift(api, hashes[n+1], n+1),
+			rightShiftNBytes(api, h, n+2)))
+	}
+	hFunc.Write(leftShift1Byte(api, leastNBytes(api, hashes[len(hashes)-1], len(hashes)-1)))
+	return hFunc.Sum()
+}
+
+// Shift lsb N Bytes to left, with 0 tails
+func NBytesLeftShift(api frontend.API, input frontend.Variable, N int) frontend.Variable {
+	var lsb frontend.Variable = 0
+	var multiplier frontend.Variable = 1
+	inputBinary := api.ToBinary(input)
+	if N >= 32 {
+		return 0
+	}
+	for i := 0; i < 8*N; i++ {
+		lsb = api.Add(lsb, api.Mul(inputBinary[i], multiplier))
+		multiplier = api.Mul(multiplier, 2)
+	}
+	multiplier = 1
+	for i := 0; i < 32-N; i++ {
+		multiplier = api.Mul(multiplier, 256)
+	}
+	return api.Mul(lsb, multiplier)
```

Comment:
what happens to input bytes that are larger than 32 Bytes? is this operation cyclic?

Reply:
I tried to find the cyclic pattern and the maximum boundary but could not find it. I tried [255,... 255] (32 bytes) + 1and [63, 255, ..., 255] (33 bytes) + 1 and [127, 255, ... 255] (33 bytes) + 1 and many other but in vain. But the thing we can guarantee is within 32 bytes the behavior is expected.

Reply:
in that case we should add a warning in the description of the function, as i think the left shift by N bytes can lead to errors

Reply:
Added descriptions in [ae3d814](https://github.com/iotaledger/trie.go/pull/1/commits/ae3d81459df2f7eb0c34a067fe239ec83df5b85e)